File: dynamic_programming/StationSchedule.java
public class StationSchedule {

    public static void main(String[] args) {
        int n = 6;
        int[][] station = {{7,9,3,4,8,4}, {8,5,6,4,5,7}};
        int[][] move = {{2,3,1,3,4}, {2,1,2,2,1}};
        int[] enter = {2,4};
        int[] exit = {3,2};
        int[][] intermediate = new int[2][n + 1];
        int[][] line = new int[2][n];
        int index = fasetestWay(station, move, enter, exit, n, intermediate, line);
        System.out.println("Fastest Way: " + intermediate[index][n]);
        printLine(line, index, n);
        //printIntermediate(intermediate, n);
    }

    public static int fasetestWay(int[][] station, int[][] move, int[] enter, int[] exit, int n,
                                  int[][] intermediate, int[][] line) {  
        int ret, prev, t0, t1;
        intermediate[0][0] = enter[0] + station[0][0];
        intermediate[1][0] = enter[1] + station[1][0];

        for(int j = 1; j < n; j++) {
            prev = j - 1;
            t0 = intermediate[0][prev] + station[0][j];
            t1 = intermediate[1][prev] + move[1][prev] + station[0][j];
            if(t0 < t1) {
                intermediate[0][j] = t0;
                line[0][j] = 0;
            } else {
                intermediate[0][j] = t1;
                line[0][j]= 1;
            }

            t0 = intermediate[1][prev] + station[1][j];
            t1 = intermediate[0][prev] + move[0][prev] + station[1][j];
            if(t0 < t1){
                intermediate[1][j] = t0;
                line[1][j] = 1;
            } else {
                intermediate[1][j] = t1;
                line[1][j] = 0;
            }
        }

        intermediate[0][n] = intermediate[0][n - 1] + exit[0];
        intermediate[1][n] = intermediate[1][n - 1] + exit[1];
        if(intermediate[0][n] <= intermediate[1][n]) {
            ret = 0;
        } else {
            ret = 1;
        }
        return ret;
    }

    public static void printLine(int[][] line, int index, int n){
        int seq = index;
        System.out.println("Station: " + n + " Line: " + (seq + 1));
        for(int i = n - 1; i >= 1; --i) {
             seq = line[seq][i];
             System.out.println("Station: " + i + " Line: " + (seq + 1));
        }
    }

    public static void printIntermediate(int[][] intermediate, int n) {
        for(int i = 0; i <= n; ++i) {
            System.out.println("" + intermediate[0][i] + "    " + intermediate[1][i]);
        }
    }
}

File: dynamic_programming/station_schedule.c
#include <stdio.h>

#define LINE_COUNT 2
#define STATION_COUNT 6

int fastest_way(int (*station)[LINE_COUNT], int (*move)[LINE_COUNT], int *enter, int *quit, int n,
                int (*line)[LINE_COUNT], int (*inter)[LINE_COUNT]) {
  int i, prev, t0, t1, ret;

  inter[0][0] = enter[0] + station[0][0];
  inter[0][1] = enter[1] + station[0][1];
  line[0][0] = 0;
  line[0][1] = 1;
  for(i = 1; i < n; ++i) {
    prev = i - 1;

    t0 = inter[prev][0] + station[i][0];
    t1 = inter[prev][1] + move[prev][1] + station[i][0];
    if(t0 < t1) {
      line[i][0] = 0;
      inter[i][0] += t0;
    } else {
      line[i][0] = 1;
      inter[i][0] += t1;
    }

    t0 = inter[prev][1] + station[i][1];
    t1 = inter[prev][0] + move[prev][0] + station[i][1];
    if(t0 < t1) {
      line[i][1] = 1;
      inter[i][1] += t0;
    } else {
      line[i][1] = 0;
      inter[i][1] += t1;
    }
  }
  inter[n][0] = inter[n - 1][0] + quit[0];
  inter[n][1] = inter[n - 1][1] + quit[1];

  if(inter[n][0] < inter[n][1]) {
    ret = 0;
  } else {
    ret = 1;
  }

  /*for(i = 0; i < n; ++i) {
    printf("%d  |  %d  |  %d  |  %d\n", inter[i][0], inter[i][1], line[i][0], line[i][1]);
  }
  printf("%d  |  %d\n", inter[i][0], inter[i][1]);*/
  return ret;
}

void print_line(int (*line)[LINE_COUNT], int n, int index) {
  int i, seq = index;
  printf("Station: %d, Line: %d\n", n, seq + 1);
  for(i = n - 1; i >= 1; --i) {
    seq = line[i][seq];
    printf("Station: %d, Line: %d\n", i, seq + 1);
  }
}

int main() {
  int station[][LINE_COUNT] = {7, 8, 9, 5, 3, 6, 4, 4, 8, 5, 4, 7};
  int move[][LINE_COUNT] = {2, 2, 3, 1, 1, 2, 3, 2, 4, 1};
  int enter[] = {2, 4};
  int quit[] = {3, 2};
  int line[STATION_COUNT][LINE_COUNT] = {0};
  int inter[STATION_COUNT + 1][LINE_COUNT] = {0};
  int index;
  index = fastest_way(station, move, enter, quit, STATION_COUNT, line, inter);
  printf("Fastest way: %d\n", inter[STATION_COUNT][index]);
  print_line(line, STATION_COUNT, index);
  return 0;
}
